File: pr1-java/client/src/main/java/pr1Java/client/javafx/thrift/clients/Window.java
package pr1Java.client.javafx.thrift.clients;

import org.apache.thrift.transport.TTransport;
import pr1Java.client.Configuration;
import pr1Java.model.User;
import pr1Java.services.thrift.ThriftServices;

public abstract class Window {
    protected TTransport connection;
    protected ThriftServices.Client services = null;
    protected User signedInUser = null;

    public void init(TTransport connection, ThriftServices.Client services, User signedInUser) {
        Configuration.logger.traceEntry("entering init with {} and {}", services, signedInUser);

        this.connection = connection;
        this.services = services;
        this.signedInUser = signedInUser;

        Configuration.logger.traceExit();
    }
}

File: pr1-cs/Client/Gtk/Thrift/Clients/Window.cs
using Gtk;
using Model;
using Services;
using Services.Reflection;
using Services.Thrift;
using Thrift.Transport;

namespace Client.Gtk.Thrift.Clients
{
    public abstract class Window
    {
        protected TTransport Connection;
        protected Services.Thrift.Services.Client Services;
        protected readonly Builder GuiElements = new Builder();
        protected User SignedInUser;
        protected global::Gtk.Window OwnedWindow;

        public virtual Window Init(TTransport connection, Services.Thrift.Services.Client services, User signedInUser)
        {
            ClientStarter.OpenWindows++;
            Connection = connection;
            Services = services;
            SignedInUser = signedInUser;

            return this;
        }

        public void Open()
        {
            OwnedWindow.Show();
        }

        protected void Close()
        {
            ClientStarter.OpenWindows--;
            if (ClientStarter.OpenWindows <= 0)
            {
                Connection.Close();
                Application.Quit();
            }
            OwnedWindow.Destroy();
        }
    }
}